### PR TITLE
Fix extra lines rendering for LinePrimitives

### PR DESCRIPTION
**Pull request by snosenzo**
**User-Facing Changes**

- Fix occasional extra lines in SceneUpdate line primitives

**Description**

Fixed instanceCount for `LINE_LOOP` primitive types. Because we add one to the `numVertices` when we have a line loop, the  number of lines in a line loop should be `numVertices - 1`, or the original number of points on the primitive. The story I added has one square defined with 4 points (not closed) and the other defined with 5 points (already closed). We then add one in calculating `numVertices` because we close the loop in `serializePositions` even if the loop is already closed. This makes the number of lines (`instanceCount`) equal to 4 for the first and 5 for the second, whereas previously it was 5 and 6. This would cause the renderable to read additional points from existing, larger `positionBuffers` that were used after an update.

Fixes: FG-4944

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/stories/SceneEntities.stories.tsx [hunk 1]
```diff
@@ -436,3 +438,425 @@ export const BasicEntities: StoryObj = {
...
+];
+
+const frame1: MessageEvent<Partial<SceneUpdate>> = {
+  topic: "scene",
+  schemaName: "foxglove.SceneUpdate",
+  sizeInBytes: 0,
+  receiveTime: { sec: 10, nsec: 0 },
+  message: lineLoopSampleData[0]!,
+};
+const frame2: MessageEvent<Partial<SceneUpdate>> = {
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/stories/SceneEntities.stories.tsx:774]: Maybe a comment somewhere as to what this is testing?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts [hunk 2]
```diff
@@ -161,7 +161,9 @@ class LinePrimitiveRenderable extends THREE.Object3D {
         this.#geometry == undefined ||
         this.#lineType !== this.#primitive.type ||
         !this.#positionBuffer ||
-        this.#positionBuffer.length < necessaryPositionBufferSize;
+        // Because `LineGeometry.setPosition` iterates through the positionBuffer to create an interleaved buffer of positions,
+        // we can't simply reuse the float32array with a larger length, we need to recreate it or we will get segments beyond the new length
+        this.#positionBuffer.length !== necessaryPositionBufferSize;
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts:166]: hm... isn't the custom value of `this.#geometry.instanceCount` (below) supposed to take care of this?

Comment (snosenzo) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts:166]: Yeah I address that in the PR description, it doesn't look like it's working for LineGeometry, but that would probably be worth mentioning here too

Comment (snosenzo) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts]: InstanceCount was working, we were calculating it wrong. I updated the code and PR description to reflect that

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts [hunk 3]
```diff
@@ -161,6 +161,8 @@ class LinePrimitiveRenderable extends THREE.Object3D {
         this.#geometry == undefined ||
         this.#lineType !== this.#primitive.type ||
         !this.#positionBuffer ||
+        // Because `LineGeometry.setPosition` iterates through the positionBuffer to create an interleaved buffer of positions,
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/primitives/RenderableLines.ts:164]: Stale comment?

**Review by jtbandes — APPROVED** [review]

**Pull request merged**